### PR TITLE
Towards supporting strings / PyBytes as waveforms

### PR DESCRIPTION
**Pull request by PierreSchnizer**
**Work in progress** 

Epics strings are views of char's. These match rather to PyBytes Objects.
Thus the user is expected to handle PyBytes Object (for Python3) and
handle any string to byte conversion.

* src/pywrapper: extending multitype to vector of strings
                 handling vectors of strings

* src/pydev_waveform.cpp : handling string arrays
         and debug strings (to be removed)

* pydevtest.db: first usage of functionality

### CODE REVIEW

File: configure/CONFIG [hunk 1]
```diff
@@ -20,8 +20,9 @@ include $(CONFIG)/CONFIG
 INSTALL_LOCATION = $(TOP)
 
 # As soon as python3 is the default configuration
-# PYTHON_CONFIG=python3-config
-PYTHON_CONFIG=python-config
+PYTHON_CONFIG=python3-config
```
Comment (klemenv) [inline configure/CONFIG:23]: This should be left intact.

File: src/pydev_waveform.cpp [hunk 2]
```diff
@@ -16,7 +16,8 @@
 
 #include <map>
 #include <string.h>
-
+#include <iostream>
```
Comment (klemenv) [inline src/pydev_waveform.cpp:19]: Even though it's C++ code, we're using printf() elsewhere in the project and we should stay consistent.

File: src/pydev_waveform.cpp [hunk 3]
```diff
@@ -29,6 +30,40 @@ struct PyDevContext {
...
+            std::cerr << "Not prepared for rec->ftvl" << rec->ftvl << "\n";
+	}
+        return false;
+    }
+
+    rec->nord = std::min(arr.size(), (size_t)rec->nelm);
+    auto val = reinterpret_cast<char*>(rec->bptr);
+    for(size_t i = 0; i < rec->nord; ++i){
+        char *cptr = &val[i * epics_string_length];
+	std::string sval = arr[i];
```
Comment (klemenv) [inline src/pydev_waveform.cpp:49]: Lines 49-59 need to be indented.

File: src/pydev_waveform.cpp [hunk 4]
```diff
@@ -29,6 +30,40 @@ struct PyDevContext {
 
 static std::map<std::string, IOSCANPVT> ioScanPvts;
 
+static bool toRecArrayValString(waveformRecord* rec, const std::vector<std::string>& arr)
+{
+    /* should be defined somewhee in epics headers */
+    const int epics_string_length = 40;
```
Comment (klemenv) [inline src/pydev_waveform.cpp:36]: From epicsTypes.h:
#define MAX_STRING_SIZE 40

File: src/pydev_waveform.cpp [hunk 5]
```diff
@@ -29,6 +30,40 @@ struct PyDevContext {
...
+    for(size_t i = 0; i < rec->nord; ++i){
+        char *cptr = &val[i * epics_string_length];
+	std::string sval = arr[i];
+	// need to foresee space for last '\0'
+	if(rec->tpro && sval.size() > epics_string_length - 1){
+	    // Indicate on console which element will be truncated where
+	    std::stringstream strm;
+	    strm << rec->name << "[" << i << "]: '";
+	    const std::string info = strm.str();
+	    std::cerr << info << sval << "' too long\n"
```
Comment (klemenv) [inline src/pydev_waveform.cpp:56]: Make the printout conditional on rec->tpro

File: src/pydev_waveform.cpp [hunk 6]
```diff
@@ -193,6 +228,9 @@ static void processRecordCb(waveformRecord* rec)
         if (rec->ftvl == menuFtypeFLOAT || rec->ftvl == menuFtypeDOUBLE) {
             std::vector<double> arr;
             ret = (PyWrapper::exec(code, (rec->tpro == 1), arr) && toRecArrayVal(rec, arr));
+	} else if (rec->ftvl == menuFtypeSTRING) {
```
Comment (klemenv) [inline src/pydev_waveform.cpp:231]: 231-233 need to be indented

File: src/pywrapper.cpp [hunk 7]
```diff
@@ -253,6 +253,7 @@ bool PyWrapper::convert(void* in_, MultiTypeValue& out)
         out.type = MultiTypeValue::Type::NONE;
         out.vi.clear();
         out.vf.clear();
+                out.type = MultiTypeValue::Type::VECTOR_FLOAT;
```
Comment (klemenv) [inline src/pywrapper.cpp:256]: Not sure this is needed? We define NONE type in line 253

File: src/pywrapper.cpp [hunk 8]
```diff
@@ -290,6 +291,17 @@ bool PyWrapper::convert(void* in_, MultiTypeValue& out)
                 out.vf.push_back(val);
                 out.type = MultiTypeValue::Type::VECTOR_FLOAT;
             }
+	    if(PyBytes_Check(el) && (out.type == MultiTypeValue::Type::NONE || out.type == MultiTypeValue::Type::VECTOR_STRING ||
```
Comment (klemenv) [inline src/pywrapper.cpp:294]: The next block needs indented

File: src/pywrapper.cpp [hunk 9]
```diff
@@ -290,6 +291,17 @@ bool PyWrapper::convert(void* in_, MultiTypeValue& out)
...
+				     out.type == MultiTypeValue::Type::VECTOR_FLOAT /* how not to end here ? */
```
Comment (klemenv) [inline src/pywrapper.cpp:295]: Is VECTOR_FLOAT appropriate here?

File: src/pywrapper.cpp [hunk 10]
```diff
@@ -290,6 +291,17 @@ bool PyWrapper::convert(void* in_, MultiTypeValue& out)
...
+				     )) {
+	        const char *cval = PyBytes_AsString(el);
+	        if (!cval && PyErr_Occurred()) {
+                    PyErr_Clear();
+                    return false;
+                }
+	        out.vs.push_back(cval);
```
Comment (klemenv) [inline src/pywrapper.cpp:302]: What happens if cval==NULL

File: src/pydev_waveform.cpp (hunk 4 above)
Comment (PierreSchnizer) [reply src/pydev_waveform.cpp]: Implemented

File: src/pywrapper.cpp (hunk 10 above)
Comment (PierreSchnizer) [reply src/pywrapper.cpp]: changed. cval is the return value. Implies PyErr_Occured to true according to documentation.
Defensive check

File: configure/CONFIG (hunk 1 above)
Comment (PierreSchnizer) [reply configure/CONFIG]: Changed back.

File: src/pydev_waveform.cpp (hunk 2 above)
Comment (PierreSchnizer) [reply src/pydev_waveform.cpp]: Changed to use printf

File: src/pywrapper.cpp (hunk 9 above)
Comment (PierreSchnizer) [reply src/pywrapper.cpp]: Nope: result of the default setting of above

**Review by klemenv — APPROVED** [review]

**Pull request merged**